### PR TITLE
Ensure that viewer handshake always happens quickly.

### LINKED ISSUE #6728: Chunking must be less aggressive until we get to a state where we can be transitioned out of pre-render
Otherwise a doc might be stalled in passive mode and cannot yet transition to visible.

### PR DESCRIPTION
**Pull request by cramforce**
This fixes #6728. THe problem was that chunking can delay the handshake with the viewer for a very long time. During this time the viewer cannot tell the doc that it is now visible–and visibility triggers chunking to go into fast mode. With this change the aggressive chunking mode is only triggered when the doc has gotten into a state where it can communicate with the viewer; and the viewer communication code itself is never delayed in a chunk.

### CODE REVIEW

File: src/runtime.js [hunk 1]
```diff
@@ -246,16 +258,26 @@ function adoptShared(global, opts, callback) {
...
+    const register = () => {
+      waitForBody(global.document, () => {
+        if (typeof fnOrStruct == 'function') {
+          fnOrStruct(global.AMP);
+        } else {
+          registerExtension(extensions, fnOrStruct.n, fnOrStruct.f, global.AMP);
+        }
+      });
+    };
+    if (typeof fnOrStruct == 'function' || fnOrStruct.p == 'high') {
```
Comment (erwinmombay) [inline src/runtime.js]: can you leave a comment about the the `.p` and how it is set

File: gulpfile.js [hunk 2]
```diff
@@ -142,7 +156,9 @@ function declareExtension(name, version, hasCss, opt_noTypeCheck,
 function buildExtensions(options) {
   for (var key in extensions) {
     var e = extensions[key];
-    buildExtension(e.name, e.version, e.hasCss, options, e.extraGlobs);
+    var o = Object.assign({}, options);
+    o = Object.assign(o, e);
```
Comment (jridgewell) [inline gulpfile.js]: Why the double assign? `var o = Object.assign({}, options, e)`

Comment (cramforce) [inline gulpfile.js]: Mostly though nesting assigns is kind of unreadable. Happy to change.

File: src/runtime.js [hunk 3]
```diff
@@ -246,16 +262,26 @@ function adoptShared(global, opts, callback) {
...
+      // "High priority" extensions do not go through chunking.
```
Comment (dvoytenko) [inline src/runtime.js]: Please make a note why `function` is considered "high" priority.

Comment (cramforce) [inline src/runtime.js]: Done.

**Pull request merged**